lib/client/reexec.go

```diff
@@ -0,0 +1,145 @@
+	// Stdout is the child process' stdout.
+	Stdout io.Writer
+	// Stderr is the child process' stderr.
+	Stderr io.Writer
+}
+
+type forkAuthCmd struct {
+	*exec.Cmd
+	signalR, signalW, killR, killW *os.File
+}
+
+// RunForkAuthenticate re-execs the current executable and waits for any of
+// the following:
+//   - The child process exits (usually in error).
+//   - The child process signals the parent that it is ready to be disowned.
+//   - The context is canceled.
+func RunForkAuthenticate(ctx context.Context, params ForkAuthenticateParams) error {
+	cmd, err := buildForkAuthenticateCommand(params)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return runForkAuthenticateChild(ctx, cmd)
+}
+
+func buildForkAuthenticateCommand(params ForkAuthenticateParams) (*forkAuthCmd, error) {
+	executable, err := getExecutable()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cmd := exec.Command(executable)
+	// Set up disown signal.
+	signalR, signalW, err := os.Pipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	killR, killW, err := os.Pipe()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	signalFd, killFd := configureReexecForOS(cmd, signalW, killR)
+
+	cmd.Args = append(cmd.Args, params.GetArgs(signalFd, killFd)...)
+	cmd.Args[0] = os.Args[0]
+	cmd.Stdin = params.Stdin
+	cmd.Stdout = params.Stdout
+	cmd.Stderr = params.Stderr
+
+	return &forkAuthCmd{
+		Cmd: cmd,
+		// Keep all pipes around to prevent them from being garbage collected.
+		signalR: signalR,
+		signalW: signalW,
+		killR:   killR,
+		killW:   killW,
+	}, nil
+}
+
+func runForkAuthenticateChild(ctx context.Context, cmd *forkAuthCmd) error {
+	defer func() {
+		// If the child is still listening, kill it. If the child successfully
+		// disowned, this will do nothing.
+		cmd.killW.Write([]byte{0x00})
+		cmd.killW.Close()
+		cmd.signalR.Close()
+	}()
+
+	disownReady := make(chan error, 1)
+	go func() {
+		// The child process will write to the pipe when it has authenticated
+		// and is ready to be disowned.
+		_, err := cmd.signalR.Read(make([]byte, 1))
+		disownReady <- err
+	}()
+
+	if err := cmd.Start(); err != nil {
+		return trace.Wrap(err)
+	}
+	for _, file := range cmd.ExtraFiles {
+		if err := file.Close(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+	select {
+	case err := <-disownReady:
+		if err == nil {
+			return trace.Wrap(cmd.Process.Release())
+		}
+
+		// Error was likely caused by the child process exiting. Wait for Wait() to
+		// return the exit status if possible.
+		childFinished := make(chan error, 1)
+		go func() {
+			childFinished <- cmd.Wait()
+		}()
+		select {
+		case waitErr := <-childFinished:
+			return trace.Wrap(waitErr)
+		case <-time.After(3 * time.Second):
+			if killErr := cmd.Process.Kill(); killErr != nil {
+				return trace.Wrap(killErr)
+			}
+			return trace.Wrap(err)
+		}
+	case <-ctx.Done():
+		if err := cmd.Process.Kill(); err != nil {
+			return trace.Wrap(err)
+		}
+		return trace.Wrap(ctx.Err())
```

Comment:
These two returns drop `cmd` without calling its `Wait`, which is liable to panic due to the finalizer guard.